### PR TITLE
feat: support convert external data to nested json

### PR DESCRIPTION
**Pull request by qinluhe**
### Feature Preview

Refactor the algorithm for converting document data to JSON, optimizing time complexity, and supporting the conversion of HTML from both Notion and Google Docs into nested JSON structures.

---

#### PR Checklist

- [ ] My code adheres to [AppFlowy's Conventions](https://docs.appflowy.io/docs/documentation/software-contributions/conventions)
- [ ] I've listed at least one issue that this PR fixes in the description above.
- [ ] I've added a test(s) to validate changes in this PR, or this PR only contains semantic changes.
- [ ] All existing tests are passing.

### CODE REVIEW

File: frontend/rust-lib/flowy-document2/src/event_handler.rs [hunk 1]
```diff
@@ -350,3 +382,34 @@ pub async fn convert_document(
       .then(|| parser.to_text_with_json(root)),
   })
 }
+
+/// Handler for converting a HTML string to a JSON string.
```
Comment (LucasXu0) [inline frontend/rust-lib/flowy-document2/src/event_handler.rs:386]: remove `HTML`

File: frontend/rust-lib/flowy-document2/src/parser/external/utils.rs [hunk 2]
```diff
@@ -0,0 +1,480 @@
...
+  if let Some(JSONResult::Block(block)) = flatten_element_to_json(node, &None, &None) {
+    return Some(block);
+  }
+
+  None
+}
+
+/// Parse plaintext to nested block
+pub fn parse_plaintext_to_nested_block(plaintext: &str) -> Option<NestedBlock> {
+  let lines: Vec<&str> = plaintext.lines().collect();
```
Comment (LucasXu0) [inline frontend/rust-lib/flowy-document2/src/parser/external/utils.rs:48]: filter, trim

File: frontend/rust-lib/flowy-document2/src/parser/external/utils.rs [hunk 3]
```diff
@@ -0,0 +1,480 @@
...
+    } else {
+      // put text into delta while child is a text node
+      let text = child
+        .value()
+        .as_text()
+        .map(|text| {
+          let text = text.text.to_string();
+
+          if text.trim().is_empty() {
+            return text.trim().to_string();
```
Comment (LucasXu0) [inline frontend/rust-lib/flowy-document2/src/parser/external/utils.rs:337]: is it equal to ''?

File: frontend/rust-lib/flowy-document2/src/parser/external/utils.rs [hunk 4]
```diff
@@ -0,0 +1,480 @@
+use crate::parser::constant::*;
+use crate::parser::parser_entities::{InsertDelta, NestedBlock};
+use scraper::node::Attrs;
+use scraper::ElementRef;
+use serde::{Deserialize, Serialize};
+use serde_json::{Map, Value};
+use std::collections::HashMap;
+
+const INLINE_TAGS: [&str; 9] = [
```
Comment (LucasXu0) [inline frontend/rust-lib/flowy-document2/src/parser/external/utils.rs:9]:
```dart
class HTMLTags {
  static const h1 = 'h1';
  static const h2 = 'h2';
  static const h3 = 'h3';
  static const orderedList = 'ol';
  static const unorderedList = 'ul';
  static const list = 'li';
  static const paragraph = 'p';
  static const image = 'img';
  static const anchor = 'a';
  static const italic = 'i';
  static const em = 'em';
  static const bold = 'b';
  static const underline = 'u';
  static const strikethrough = 's';
  static const del = 'del';
  static const strong = 'strong';
  static const checkbox = 'input';
  static const span = 'span';
  static const code = 'code';
  static const blockQuote = 'blockquote';
  static const div = 'div';
  static const divider = 'hr';
  static const table = 'table';
  static const tableRow = 'tr';
  static const tableheader = "th";
  static const tabledata = "td";
  static const section = 'section';
  static const font = 'font';
  static const mark = 'mark';

  static List<String> formattingElements = [
    HTMLTags.anchor,
    HTMLTags.italic,
    HTMLTags.em,
    HTMLTags.bold,
    HTMLTags.underline,
    HTMLTags.del,
    HTMLTags.strong,
    HTMLTags.span,
    HTMLTags.code,
    HTMLTags.strikethrough,
    HTMLTags.font,
    HTMLTags.mark,
  ];
}
```

Comment (LucasXu0) [reply frontend/rust-lib/flowy-document2/src/parser/external/utils.rs:9]: These are the inline tags we're using.

File: frontend/rust-lib/event-integration/tests/document/local_test/test.rs [hunk 5]
```diff
@@ -125,7 +127,7 @@ async fn apply_text_delta_test() {
 macro_rules! generate_convert_document_test_cases {
```
Comment (appflowy) [inline frontend/rust-lib/event-integration/tests/document/local_test/test.rs:127]: Nice try

File: frontend/rust-lib/flowy-document2/src/parser/parser_entities.rs [hunk 6]
```diff
@@ -479,3 +454,51 @@ impl NestedBlock {
...
+
+#[derive(PartialEq, Eq, Debug, ProtoBuf_Enum, Clone, Default)]
+pub enum InputType {
+  #[default]
+  Html = 0,
+  PlainText = 1,
+}
+
+#[derive(Default, ProtoBuf, Debug)]
+pub struct ConvertDataToJsonPayloadPB {
```
Comment (appflowy) [inline frontend/rust-lib/flowy-document2/src/parser/parser_entities.rs:471]: You can try this.
image
image

File: frontend/rust-lib/event-integration/tests/document/local_test/test.rs [hunk 7]
```diff
@@ -153,3 +155,36 @@ async fn convert_document_test() {
...
+  let test = DocumentEventTest::new().await;
+  let _ = test.create_document().await;
+
+  let html = r#"<p>Hello</p><p> World!</p>"#;
+  let payload = ConvertDataToJsonPayloadPB {
+    data: html.to_string(),
+    input_type: InputType::Html,
+  };
+  let result = test.convert_data_to_json(payload).await;
+  let expect_json = json!({ "type": "page", "data": {}, "children": [{ "type": "paragraph", "children": [], "data": { "delta": [{ "insert": "Hello" }] } }, { "type": "paragraph", "children": [], "data": { "delta": [{ "insert": " World!" }] } }] });
```
Comment (appflowy) [inline frontend/rust-lib/event-integration/tests/document/local_test/test.rs:174]: I am curious about why this json! is not auto format. You can try `cargo fmt --all` in the rust-lib directory

File: frontend/rust-lib/flowy-document2/src/event_handler.rs [hunk 8]
```diff
@@ -350,3 +382,34 @@ pub async fn convert_document(
...
+/// let test = DocumentEventTest::new().await;
+/// let payload = ConvertDataToJsonPayloadPB {
+///  data: "<p>Hello</p><p> World!</p>".to_string(),
+///  input_type: InputTypePB::Html,
+/// };
+/// let result: ConvertDataToJsonResponsePB = test.convert_data_to_json(payload).await;
+/// let expect_json = json!({ "type": "page", "data": {}, "children": [{ "type": "paragraph", "children": [], "data": { "delta": [{ "insert": "Hello" }] } }, { "type": "paragraph", "children": [], "data": { "delta": [{ "insert": " World!" }] } }] });
+/// assert!(serde_json::from_str::<NestedBlock>(&result.json).unwrap().eq(&serde_json::from_value::<NestedBlock>(expect_json).unwrap()));
+/// ```
+pub(crate) async fn convert_data_to_json_handler(
```
Comment (appflowy) [inline frontend/rust-lib/flowy-document2/src/event_handler.rs:401]:
```suggestion
pub(crate) async fn convert_data_to_json_handler(
    data: AFPluginData<ConvertDataToJsonPayloadPB>,
) -> DataResult<ConvertDataToJsonResponsePB, FlowyError> {
    let payload: ConvertDataToJsonParams = data.into_inner().try_into()?;
    let parser = ExternalDataToNestedJSONParser::new(payload.data, payload.input_type);

    let result = match parser.to_nested_block() {
        Some(res) => serde_json::to_string(&res)?,
        None => "".to_string(),
    };

    data_result_ok(ConvertDataToJsonResponsePB { json: result })
}
```

File: frontend/rust-lib/flowy-document2/src/event_handler.rs [hunk 9]
```diff
@@ -350,3 +382,34 @@ pub async fn convert_document(
...
+  data: AFPluginData<ConvertDataToJsonPayloadPB>,
+) -> DataResult<ConvertDataToJsonResponsePB, FlowyError> {
+  let payload: ConvertDataToJsonParams = data.into_inner().try_into()?;
+  let parser = ExternalDataToNestedJSONParser::new(payload.data, payload.input_type);
+  if let Some(result) = parser.to_nested_block() {
```
Comment (appflowy) [inline frontend/rust-lib/flowy-document2/src/event_handler.rs:406]: using match
```rust
pub(crate) async fn convert_data_to_json_handler(
  data: AFPluginData<ConvertDataToJsonPayloadPB>,
) -> DataResult<ConvertDataToJsonResponsePB, FlowyError> {
  let payload: ConvertDataToJsonParams = data.into_inner().try_into()?;
  let parser = ExternalDataToNestedJSONParser::new(payload.data, payload.input_type);
  if let Some(result) = parser.to_nested_block() {
    return data_result_ok(ConvertDataToJsonResponsePB {
      json: serde_json::to_string(&result).unwrap_or_default(),
    });
  }

  data_result_ok(ConvertDataToJsonResponsePB {
    json: "".to_string(),
  })
}
```

File: frontend/rust-lib/flowy-document2/src/parser/parser_entities.rs [hunk 10]
```diff
@@ -68,7 +68,7 @@ pub struct ConvertDocumentPayloadPB {
   pub range: Option<RangePB>,
 
   #[pb(index = 3)]
-  pub export_types: ExportTypePB,
+  pub export_types: ParseTypePB,
```
Comment (appflowy) [inline frontend/rust-lib/flowy-document2/src/parser/parser_entities.rs:71]: rename export_type to parse_type?

File: frontend/rust-lib/flowy-error/src/code.rs [hunk 11]
```diff
@@ -259,6 +259,9 @@ pub enum ErrorCode {
 
   #[error("Not support yet")]
   NotSupportYet = 85,
+
+  #[error("Invalid data")]
+  InvalidData = 86,
```
Comment (appflowy) [inline frontend/rust-lib/flowy-error/src/code.rs:264]: You can use InvalidParams

File: frontend/rust-lib/flowy-document2/src/parser/external/parser.rs [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+    Self {
+      external_data: data,
+      input_type,
+    }
+  }
+
+  /// Format to nested block.
+  ///
+  /// Example:
+  /// ```json
```
Comment (appflowy) [inline frontend/rust-lib/flowy-document2/src/parser/external/parser.rs:25]: I don't get it. The JSON here seems vague

File: frontend/rust-lib/flowy-document2/src/event_map.rs [hunk 13]
```diff
@@ -79,48 +82,17 @@ pub enum DocumentEvent {
...
-  ///       index: 0,
-  ///       length: 5,
-  ///     },
-  ///     end: SelectionPB {
-  ///       block_id: "2".to_string(),
-  ///       index: 5,
-  ///       length: 7,
-  ///     }
-  ///   }),
-  ///   export_types: ConvertTypePB {
```
Comment (appflowy) [inline frontend/rust-lib/flowy-document2/src/event_map.rs:109]: what is the `ConvertTypePB`

File: frontend/rust-lib/flowy-document2/src/event_handler.rs [hunk 14]
```diff
@@ -309,16 +311,46 @@ impl From<(&Vec<BlockEvent>, bool)> for DocEventPB {
...
+///       index: 0,
+///       length: 5,
+///     },
+///     end: SelectionPB {
+///       block_id: "2".to_string(),
+///       index: 5,
+///       length: 7,
+///     }
+///   }),
+///   export_types: ConvertTypePB {
```
Comment (appflowy) [inline frontend/rust-lib/flowy-document2/src/event_handler.rs:341]: what is the ConvertTypePB

Comment (qinluhe) [reply frontend/rust-lib/flowy-document2/src/event_handler.rs:341]: It's ParseTypePB, i forgot to update this

File: frontend/rust-lib/flowy-error/src/code.rs (hunk 11 above)
Comment (qinluhe) [reply frontend/rust-lib/flowy-error/src/code.rs:264]: i will remove it

File: frontend/rust-lib/flowy-document2/src/parser/external/parser.rs (hunk 12 above)
Comment (qinluhe) [reply frontend/rust-lib/flowy-document2/src/parser/external/parser.rs:27]: I modify the comment
```
  /// Format to nested block.
  ///
  /// Example:
  /// - input html: <p><strong>Hello</strong></p><p> World!</p>
  /// - output json:
  /// { "type": "page", "data": {}, "children": [{ "type": "paragraph", "children": [], "data": { "delta": [{ "insert": "Hello", attributes: { "bold": true } }] } }, { "type": "paragraph", "children": [], "data": { "delta": [{ "insert": " World!", attributes: null }] } }] }
```

**Pull request merged**